securedrop/models.py

```diff
@@ -784,13 +781,89 @@ def to_json(self, all_info: bool = True) -> Dict[str, Any]:
 
         return json_user
 
+    def is_deleted_user(self) -> bool:
+        """Is this the special "deleted" user managed by the system?"""
+        return self.username == "deleted"
+
+    def delete(self) -> None:
+        """delete a journalist, migrating some data over to the "deleted" journalist
+
+        Callers must commit the session themselves
+        """
+        deleted = DeletedJournalist.get()
+        # All replies should be reassociated with the "deleted" journalist
+        for reply in Reply.query.filter_by(journalist_id=self.id).all():
+            reply.journalist_id = deleted.id
+            db.session.add(reply)
+
+        # For seen indicators, we need to make sure one doesn't already exist
+        # otherwise it'll hit a unique key conflict
+        already_seen_files = {
+            file.file_id for file in
+            SeenFile.query.filter_by(journalist_id=deleted.id).all()}
+        for file in SeenFile.query.filter_by(journalist_id=self.id).all():
+            if file.file_id in already_seen_files:
+                db.session.delete(file)
+            else:
+                file.journalist_id = deleted.id
+                db.session.add(file)
+
+        already_seen_messages = {
+            message.message_id for message in
+            SeenMessage.query.filter_by(journalist_id=deleted.id).all()}
+        for message in SeenMessage.query.filter_by(journalist_id=self.id).all():
+            if message.message_id in already_seen_messages:
+                db.session.delete(message)
+            else:
+                message.journalist_id = deleted.id
+                db.session.add(message)
+
+        already_seen_replies = {
+            reply.reply_id for reply in
+            SeenReply.query.filter_by(journalist_id=deleted.id).all()}
+        for reply in SeenReply.query.filter_by(journalist_id=self.id).all():
+            if reply.reply_id in already_seen_replies:
+                db.session.delete(reply)
+            else:
+                reply.journalist_id = deleted.id
+                db.session.add(reply)
+
+        # For the rest of the associated data we rely on cascading deletions
+        db.session.delete(self)
+
+
+class DeletedJournalist(Journalist):
+    """A system user that represents deleted journalists for referential integrity"""
+    def __init__(self) -> None:
+        # password field is unused (we override set_password below), plus
+        # it's too short to be valid anyway.
+        super().__init__(username='deleted', password='unused')
```

Comment:
A valid OTP secret is generated for the `deleted` user - attempting to log in with said credentials fails with the following raceback:
```
ValueError

ValueError: Should never happen: pw_salt is none for legacy Journalist 4
Traceback (most recent call last)

    File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/flask/app.py", line 2309, in __call__

    return self.wsgi_app(environ, start_response)

    File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/flask/app.py", line 2295, in wsgi_app

    response = self.handle_exception(e)

    File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/flask/app.py", line 1741, in handle_exception

    reraise(exc_type, exc_value, tb)

    File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/flask/_compat.py", line 35, in reraise

    raise value

    File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/flask/app.py", line 2292, in wsgi_app

    response = self.full_dispatch_request()

    File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/flask/app.py", line 1815, in full_dispatch_request

    rv = self.handle_user_exception(e)

    File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/flask/app.py", line 1718, in handle_user_exception

    reraise(exc_type, exc_value, tb)

    File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/flask/_compat.py", line 35, in reraise

    raise value

    File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/flask/app.py", line 1813, in full_dispatch_request

    rv = self.dispatch_request()

    File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/flask/app.py", line 1799, in dispatch_request

    return self.view_functions[rule.endpoint](**req.view_args)

    File "/home/user/securedrop/securedrop/journalist_app/main.py", line 30, in login

    user = validate_user(request.form['username'],

    File "/home/user/securedrop/securedrop/journalist_app/utils.py", line 97, in validate_user

    return Journalist.login(username, password, token)

    File "/home/user/securedrop/securedrop/models.py", line 730, in login

    if not user.valid_password(password):

    File "/home/user/securedrop/securedrop/models.py", line 571, in valid_password

    raise ValueError(

    ValueError: Should never happen: pw_salt is none for legacy Journalist 4
```
Login should fail, but not this explosively! There was existing code to bail out of logins for invalid usernames, of which `deleted` is currently the only one, but it's not getting hit because it assumes that the invalid username and UUID are the same. 

I'm kindof coming around to the idea of a "login_enabled" field or something in the journalist table, which would be checked first when validating user logins. This will probably have to be added anyway in the future to support account locking, but it could simplify this PR now, and allow for the removal of that INVALID_USERNAMES array in favour of a db query.

Reply:
Also, setting/forgetting a valid diceware password for the deleted user, and dropping the overriden set_password method, would both rule out the exception above and clear the new LGTM alert. I recognise that it's an intentional, not accidental, change, but the side effects of overriding the method seem undefined, which is a worry.

Reply:
I'm just leery of something like that throwing an undhandled exception, but the point about migration woes is well-taken. Maybe we don't even need the `login_enabled` field just yet: if you had a `Journalist.can_log_in()` that just returned true you could override that in `DeletedJournalist` to return false without falling foul of constructor overrides. If we ever did add account locking (and said field) it would still be useful, the `Journalist` version of the method would probably just query the db.

Reply:
Creating a PassphraseGenerator in the migration step seems to work with minimal copy and paste, so I pushed changes that now set a real passphrase/totp secret for "deleted", and no longer override "set_password". I also fixed the pre-existing tests for the deleted journalist login to work with the new style of having a real UUID.

> if you had a `Journalist.can_log_in()` that just returned true you could override that in `DeletedJournalist` to return false without falling foul of constructor overrides

This wouldn't work because on the login form when we query by username, we get an instance of `Journalist` back, not `DeletedJournalist`. Now `DeletedJournalist` is only used to bypass the username validation in check_username_acceptable.

Now that we no longer need it to override password handling, we could just get rid of the class and do something like:
```
deleted = Journalist(username="placeholder", password=...)
deleted.username = "deleted"
```
to bypass the constructor validation. And `DeletedJournalist.get()` would become `Journalist.get_deleted()`. What do you think?

Reply:
SGTM - will hold off on further review for now, lemme know when it's updated.

Reply:
@zenmonkeykstop done!